README.rst
Pilbox
======

.. image:: https://badge.fury.io/py/pilbox.svg
    :target: https://pypi.python.org/pypi/pilbox
    :alt: PyPI Version

.. image:: https://travis-ci.org/agschwender/pilbox.svg?branch=master
    :target: https://travis-ci.org/agschwender/pilbox
    :alt: Build Status

.. image:: https://coveralls.io/repos/agschwender/pilbox/badge.svg
    :target: https://coveralls.io/r/agschwender/pilbox
    :alt: Test Coverage

.. image:: https://landscape.io/github/agschwender/pilbox/master/landscape.svg?style=flat
    :target: https://landscape.io/github/agschwender/pilbox/master
    :alt: Code Health


Pilbox is an image processing application server built on Python's
`Tornado web framework <http://www.tornadoweb.org/en/stable/>`_ using
the `Python Imaging Library
(Pillow) <https://pypi.python.org/pypi/Pillow/>`_. It is not
intended to be the primary source of images, but instead acts as a proxy
which requests images and resizes them as desired.

Setup
=====

Dependencies
------------

-  >= `Python 2.7 <http://www.python.org/download/>`_
-  `Pillow 2.8.1 <https://pypi.python.org/pypi/Pillow/2.8.1>`_
-  `Tornado 4.0.2 <https://pypi.python.org/pypi/tornado/4.0.2>`_
-  `OpenCV 2.x <http://opencv.org/>`_ (optional)
-  `PycURL 7.x <http://pycurl.sourceforge.net/>`_ (optional, but
   recommended; required for proxy requests and requests over TLS)
-  Image Libraries: libjpeg-dev, libfreetype6-dev, libwebp-dev,
   zlib1g-dev, liblcms2-dev

Pilbox highly recommends installing ``libcurl`` and ``pycurl`` in order
to get better HTTP request performance as well as additional features
such as proxy requests and requests over TLS. Installed versions of
``libcurl`` should be a minimum of ``7.21.1`` and ``pycurl`` should be a
minimum of ``7.18.2``. Furthermore, it is recommended that the
``libcurl`` installation be built with asynchronous DNS resolver
(threaded or c-ares), otherwise it may encounter various problems with
request timeouts (for more information, see `CURLOPT_CONNECTTIMEOUT_MS <http://curl.haxx.se/libcurl/c/curl_easy_setopt.html#CURLOPTCONNECTTIMEOUTMS>`_
and comments in `curl_httpclient.py <https://github.com/tornadoweb/tornado/blob/master/tornado/curl_httpclient.py>`_)

Install
-------

Pilbox can be installed with pip

::

    $ pip install pilbox

Or easy_install

::

    $ easy_install pilbox

Or from source

::

    $ git clone https://github.com/agschwender/pilbox.git

Packaged with Pilbox is a `Vagrant <http://www.vagrantup.com/>`_
configuration file which installs all necessary dependencies on a
virtual box using `Ansible <http://www.ansibleworks.com/>`_. See the
`Vagrant documentation <http://docs.vagrantup.com/v2/installation/>`_
and the `Ansible
documentation <http://www.ansibleworks.com/docs/gettingstarted.html#getting-ansible>`_
for installation instructions. Once installed, the following will start
and provision a virtual machine.

::

    $ vagrant up
    $ vagrant provision

To access the virtual machine itself, simply...

::

    $ vagrant ssh

When running via Vagrant, the application is automatically started on
port 8888 on 192.168.100.100, i.e.

::

    http://192.168.100.100:8888/


Running
=======

To run the application, issue the following command

::

    $ python -m pilbox.app

By default, this will run the application on port 8888 and can be
accessed by visiting:

::

    http://localhost:8888/

To see a list of all available options, run

::

    $ python -m pilbox.app --help
    Usage: pilbox/app.py [OPTIONS]

    Options:

      --allowed_hosts            list of allowed hosts (default [])
      --allowed_operations       list of allowed operations (default [])
      --background               default hexadecimal bg color (RGB or ARGB)
      --ca_certs                 filename of CA certificates in PEM format
      --client_key               client key
      --client_name              client name
      --config                   path to configuration file
      --content_type_from_image  override content type using image mime type
      --debug                    run in debug mode
      --expand                   default to expand when rotating
      --filter                   default filter to use when resizing
      --help                     show this help information
      --implicit_base_url        prepend protocol/host to url paths
      --max_requests             max concurrent requests (default 40)
      --operation                default operation to perform
      --optimize                 default to optimize when saving
      --port                     run on the given port (default 8888)
      --position                 default cropping position
      --preserve_exif            default behavior for Exif information
      --progressive              default to progressive when saving
      --proxy_host               proxy hostname
      --proxy_port               proxy port
      --quality                  default jpeg quality, 1-99 or keep
      --remove_alpha             remove alpha channel (default False)
      --retain                   default adaptive retain percent, 1-99
      --timeout                  timeout of requests in seconds (default 10)
      --validate_cert            validate certificates (default True)


Calling
=======

To use the image processing service, include the application url as you
would any other image. E.g. this image url

::

    <img src="http://i.imgur.com/zZ8XmBA.jpg" />

Would be replaced with this image url

::

    <img src="http://localhost:8888/?url=http%3A%2F%2Fi.imgur.com%2FzZ8XmBA.jpg&w=300&h=300&mode=crop" />

This will request the image served at the supplied url and resize it to
``300x300`` using the ``crop`` mode. The below is the list of parameters
that can be supplied to the service.

General Parameters
------------------

-  *url*: The url of the image to be resized
-  *op*: The operation to perform: noop, region, resize (default), rotate

   -  *noop*: No operation is performed, image is returned as it is
      received
   -  *region*: Select a sub-region from the image
   -  *resize*: Resize the image
   -  *rotate*: Rotate the image

-  *fmt*: The output format to save as, defaults to the source format

   -  *gif*: Save as GIF
   -  *jpeg*: Save as JPEG
   -  *png*: Save as PNG
   -  *webp*: Save as WebP
   -  *tiff*: Save as TIFF

-  *opt*: The output should be optimized, only relevant to JPEGs and PNGs
-  *exif*: Keep original `Exif <http://en.wikipedia.org/wiki/Exchangeable_image_file_format>`_
   data in the processed image, only relevant for JPEG
-  *prog*: Enable progressive output, only relevant to JPEGs
-  *q*: The quality, (1-99) or keep, used to save the image, only relevant
   to JPEGs

Resize Parameters
-----------------

-  *w*: The desired width of the image
-  *h*: The desired height of the image
-  *mode*: The resizing method: adapt, clip, crop (default), fill and scale

   -  *adapt*: Resize using crop if the resized image retains a supplied
      percentage of the original image; otherwise fill
   -  *clip*: Resize to fit within the desired region, keeping aspect
      ratio
   -  *crop*: Resize so one dimension fits within region, center, cut
      remaining
   -  *fill*: Fills the clipped space with a background color
   -  *scale*: Resize to fit within the desired region, ignoring aspect
      ratio

-  *bg*: Background color used with fill mode (RGB or ARGB)

   -  *RGB*: 3- or 6-digit hexadecimal number
   -  *ARGB*: 4- or 8-digit hexadecimal number, only relevant for PNG
      images

-  *filter*: The filtering algorithm used for resizing

   -  *nearest*: Fastest, but often images appear pixelated
   -  *bilinear*: Faster, can produce acceptable results
   -  *bicubic*: Fast, can produce acceptable results
   -  *antialias*: Slower, produces the best results

-  *pos*: The crop position

   -  *top-left*: Crop from the top left
   -  *top*: Crop from the top center
   -  *top-right*: Crop from the top right
   -  *left*: Crop from the center left
   -  *center*: Crop from the center
   -  *right*: Crop from the center right
   -  *bottom-left*: Crop from the bottom left
   -  *bottom*: Crop from the bottom center
   -  *bottom-right*: Crop from the bottom right
   -  *face*: Identify faces and crop from the midpoint of their
      position(s)
   -  *x,y*: Custom center point position ratio, e.g. 0.0,0.75

-  *retain*: The minimum percentage (1-99) of the original image that
   must still be visible in the resized image in order to use crop mode


Region Parameters
-----------------

-  *rect*: The region as x,y,w,h; x,y: top-left position, w,h:
   width/height of region

Rotate Parameters
-----------------

-  *deg*: The desired rotation angle degrees

   - *0-359*: The number of degrees to rotate (clockwise)
   - *auto*: Auto rotation based on Exif orientation, only relevant to JPEGs

-  *expand*: Expand the size to include the full rotated image

Security-related Parameters
---------------------------

-  *client*: The client name
-  *sig*: The signature

The ``url`` parameter is always required as it dictates the image that
will be manipulated. ``op`` is optional and defaults to ``resize``. It
also supports a comma separated list of operations, where each operation
is applied in the order that it appears in the list. Depending on the
operation, additional parameters are required. All image manipulation
requests accept ``exif``, ``fmt``, ``opt``, ``prog`` and ``q``. ``exif``
is optional and default to ``0`` (not preserved). ``fmt`` is optional
and defaults to the source image format. ``opt`` is optional and
defaults to ``0`` (disabled). ``prog`` is optional and default to ``0``
(disabled). ``q`` is optional and defaults to ``90``. To ensure
security, all requests also support, ``client`` and ``sig``. ``client``
is required only if the ``client_name`` is defined within the
configuration file. Likewise, ``sig`` is required only if the
``client_key`` is defined within the configuration file. See the
`Signing`_ section for details on how to generate the signature.

For resizing, either the ``w`` or ``h`` parameter is required. If only
one dimension is specified, the application will determine the other
dimension using the aspect ratio. ``mode`` is optional and defaults to
``crop``. ``filter`` is optional and defaults to ``antialias``. ``bg``
is optional and defaults to ``fff``. ``pos`` is optional and defaults to
``center``. ``retain`` is optional and defaults to ``75``.

For region sub-selection, ``rect`` is required. For rotating, ``deg`` is
required. ``expand`` is optional and defaults to ``0`` (disabled). It is
recommended that this feature not be used as it typically does not
produce high quality images.

Note, all built-in defaults can be overridden by setting them in the
configuration file. See the `Configuration`_ section
for more details.

Examples
========

The following images show the various resizing modes in action for an
original image size of ``640x428`` that is being resized to ``500x400``.

Adapt
-----

The adaptive resize mode combines both `crop`_ and `fill`_ resize modes
to ensure that the image always matches the requested size and a minimum
percentage of the image is always visible. Adaptive resizing will first
calculate how much of the image will be retained if crop is used. Then,
if that percentage is equal to or above the requested minimum retained
percentage, crop mode will be used. If it is not, fill will be used. The
first figure uses a ``retain`` value of ``80`` to illustrate the
adaptive crop behavior.

.. figure:: https://github.com/agschwender/pilbox/raw/master/pilbox/test/data/expected/example-500x400-mode=adapt-retain=80.jpg
     :align: center
     :alt: Adaptive cropped image

Whereas the second figure requires a minimum of ``99`` to illustrate the
adaptive fill behavior

.. figure:: https://github.com/agschwender/pilbox/raw/master/pilbox/test/data/expected/example-500x400-mode=adapt-background=ccc-retain=99.jpg
     :align: center
     :alt: Adaptive filled image

Clip
----

The image is resized to fit within a ``500x400`` box, maintaining aspect
ratio and producing an image that is ``500x334``. Clipping is useful
when no portion of the image can be lost and it is acceptable that the
image not be exactly the supplied dimensions, but merely fit within the
dimensions.

.. figure:: https://github.com/agschwender/pilbox/raw/master/pilbox/test/data/expected/example-500x400-clip.jpg
     :align: center
     :alt: Clipped image

Crop
----

The image is resized so that one dimension fits within the ``500x400``
box. It is then centered and the excess is cut from the image. Cropping
is useful when the position of the subject is known and the image must
be exactly the supplied size.

.. figure:: https://github.com/agschwender/pilbox/raw/master/pilbox/test/data/expected/example-500x400-crop.jpg
     :align: center
     :alt: Cropped image


Fill
----

Similar to clip, fill resizes the image to fit within a ``500x400`` box.
Once clipped, the image is centered within the box and all left over
space is filled with the supplied background color. Filling is useful
when no portion of the image can be lost and it must be exactly the
supplied size.

.. figure:: https://github.com/agschwender/pilbox/raw/master/pilbox/test/data/expected/example-500x400-fill-ccc.jpg
    :align: center
    :alt: Filled image


Scale
-----

The image is clipped to fit within the ``500x400`` box and then
stretched to fill the excess space. Scaling is often not useful in
production environments as it generally produces poor quality images.
This mode is largely included for completeness.

.. figure:: https://github.com/agschwender/pilbox/raw/master/pilbox/test/data/expected/example-500x400-scale.jpg
    :align: center
    :alt: Scale image


Testing
=======

To run all tests, issue the following command

::

    $ python -m pilbox.test.runtests

To run individual tests, simply indicate the test to be run, e.g.

::

    $ python -m pilbox.test.runtests pilbox.test.signature_test

Signing
=======

In order to secure requests so that unknown third parties cannot easily
use the resize service, the application can require that requests
provide a signature. To enable this feature, set the ``client_key``
option. The signature is a hexadecimal digest generated from the client
key and the query string using the HMAC-SHA1 message authentication code
(MAC) algorithm. The below python code provides an example
implementation.

::

    import hashlib
    import hmac

    def derive_signature(key, qs):
        m = hmac.new(key, None, hashlib.sha1)
        m.update(qs)
        return m.hexdigest()

The signature is passed to the application by appending the ``sig``
parameter to the query string; e.g.
``x=1&y=2&z=3&sig=c9516346abf62876b6345817dba2f9a0c797ef26``. Note, the
application does not include the leading question mark when verifying
the supplied signature. To verify your signature implementation, see the
``pilbox.signature`` command described in the `Tools`_ section.

Configuration
=============

All options that can be supplied to the application via the command
line, can also be specified in the configuration file. Configuration
files are simply python files that define the options as variables. The
below is an example configuration.

::

    # General settings
    port = 8888

    # Set client name and key if the application requires signed requests. The
    # client must sign the request using the client_key, see README for
    # instructions.
    client_name = "sample"
    client_key = "3NdajqH8mBLokepU4I2Bh6KK84GUf1lzjnuTdskY"

    # Set the allowed hosts as an alternative to signed requests. Only those
    # images which are served from the following hosts will be requested.
    allowed_hosts = ["localhost"]

    # Request-related settings
    max_requests = 50
    timeout = 7.5

    # Set default resizing options
    background = "ccc"
    filter = "bilinear"
    mode = "crop"
    position = "top"

    # Set default rotating options
    expand = False

    # Set default saving options
    format = None
    optimize = 1
    quality = "90"

Tools
=====

To verify that your client application is generating correct signatures,
use the signature command.

::

    $ python -m pilbox.signature --key=abcdef "x=1&y=2&z=3"
    Query String: x=1&y=2&z=3
    Signature: c9516346abf62876b6345817dba2f9a0c797ef26
    Signed Query String: x=1&y=2&z=3&sig=c9516346abf62876b6345817dba2f9a0c797ef26

The application allows the use of the resize functionality via the
command line.

::

    $ python -m pilbox.image --width=300 --height=300 http://i.imgur.com/zZ8XmBA.jpg > /tmp/foo.jpg

If a new mode is added or a modification was made to the libraries that
would change the current expected output for tests, run the generate
test command to regenerate the expected output for the test cases.

::

    $ python -m pilbox.test.genexpected

Deploying
=========

The application itself does not include any caching. It is recommended
that the application run behind a CDN for larger applications or behind
varnish for smaller ones.

Defaults for the application have been optimized for quality rather than
performance. If you wish to get higher performance out of the
application, it is recommended you use a less computationally expensive
filtering algorithm and a lower JPEG quality. For example, add the
following to the configuration.

::

    # Set default resizing options
    filter = "bicubic"
    quality = 75

Extension
=========

While it is generally recommended to use Pilbox as a standalone server, it can also be used as a library. To extend from it and build a custom image processing server, use the following example.

::

    #!/usr/bin/env python

    import tornado.gen

    from pilbox.app import PilboxApplication, ImageHandler, \
        start_server, parse_command_line


    class CustomApplication(PilboxApplication):
        def get_handlers(self):
            return [(r"/(\d+)x(\d+)/(.+)", CustomImageHandler)]


    class CustomImageHandler(ImageHandler):
        def prepare(self):
            self.args = self.request.arguments.copy()

        @tornado.gen.coroutine
        def get(self, w, h, url):
            self.args.update(dict(w=w, h=h, url=url))

            self.validate_request()
            resp = yield self.fetch_image()
            self.render_image(resp)

        def get_argument(self, name, default=None):
            return self.args.get(name, default)


    if __name__ == "__main__":
        parse_command_line()
        start_server(CustomApplication())


Changelog
=========

-  0.1: Image resizing fit
-  0.1.1: Image cropping
-  0.1.2: Image scaling
-  0.2: Configuration integration
-  0.3: Signature generation
-  0.3.1: Signature command-line tool
-  0.4: Image resize command-line tool
-  0.5: Facial recognition cropping
-  0.6: Fill resizing mode
-  0.7: Resize using crop position
-  0.7.1: Resize using a single dimension, maintaining aspect ratio
-  0.7.2: Added filter and quality options
-  0.7.3: Support python 3
-  0.7.4: Fixed cli for image generation
-  0.7.5: Write output in 16K blocks
-  0.8: Added support for ARGB (alpha-channel)
-  0.8.1: Increased max clients and write block sizes
-  0.8.2: Added configuration for max clients and timeout
-  0.8.3: Only allow http and https protocols
-  0.8.4: Added support for WebP
-  0.8.5: Added format option and configuration overrides for mode and
   format
-  0.8.6: Added custom position support
-  0.9: Added rotate operation
-  0.9.1: Added sub-region selection operation
-  0.9.4: Added Pilbox as a PyPI package
-  0.9.10: Converted README to reStructuredText
-  0.9.14: Added Sphinx docs
-  0.9.15: Added implicit base url to configuration
-  0.9.16: Added validate cert to configuration
-  0.9.17: Added support for GIF format
-  0.9.18: Fix for travis builds on python 2.6 and 3.3
-  0.9.19: Validate cert fix
-  0.9.20: Added optimize option
-  0.9.21: Added console script entry point
-  1.0.0: Modified for easier library usage
-  1.0.1: Added allowed operations and default operation
-  1.0.2: Modified to allow override of http content type
-  1.0.3: Safely catch image save errors
-  1.0.4: Added progressive option
-  1.1.0: Proxy server support

CHANGES.txt
  * 0.1: Image resizing fit
  * 0.1.1: Image cropping
  * 0.1.2: Image scaling
  * 0.2: Configuration integration
  * 0.3: Signature generation
  * 0.3.1: Signature command-line tool
  * 0.4: Image resize command-line tool
  * 0.5: Facial recognition cropping
  * 0.6: Fill resizing mode
  * 0.7: Resize using crop position
  * 0.7.1: Resize using a single dimension, maintaining aspect ratio
  * 0.7.2: Added filter and quality options
  * 0.7.3: Support python 3
  * 0.7.4: Fixed cli for image generation
  * 0.7.5: Write output in 16K blocks
  * 0.8: Added support for ARGB (alpha-channel)
  * 0.8.1: Increased max clients and write block sizes
  * 0.8.2: Added configuration for max clients and timeout
  * 0.8.3: Only allow http and https protocols
  * 0.8.4: Added support for WebP
  * 0.8.5: Added format option and configuration overrides for mode and format
  * 0.8.6: Added custom position support
  * 0.9: Added rotate operation
  * 0.9.1: Added sub-region selection operation
  * 0.9.4: Added Pilbox as a PyPI package
  * 0.9.10: Converted README to reStructuredText
  * 0.9.14: Added Sphinx docs
  * 0.9.15: Added implicit base url
  * 0.9.16: Added validate cert to configuration
  * 0.9.17: Added support for GIF format
  * 0.9.18: Fix for travis builds on python 2.6 and 3.3
  * 0.9.19: Validate cert fix
  * 0.9.20: Added optimize option
  * 0.9.21: Added console script entry point
  * 1.0.0: Modified for easier library usage
  * 1.0.1: Added allowed operations and default operation
  * 1.0.2: Modified to allow override of http content type
  * 1.0.3: Safely catch image save errors
  * 1.0.4: Added progressive option
  * 1.1.0: Proxy server support
  * 1.1.1: Added JPEG auto rotation based on Exif orientation
  * 1.1.2: Added keep JPEG quality option and set JPEG subsampling to keep
  * 1.1.3: Fix auto rotation on JPEG with missing Exif data
  * 1.1.4: Exception handling around invalid Exif data
  * 1.1.5: Fix image requests without content types
  * 1.1.6: Support custom applications that need command line arguments
  * 1.1.7: Support adapt resize mode
  * 1.1.8: Add preserve Exif flag
  * 1.1.9: Increase Pillow version to 2.8.1
  * 1.1.10: Add ca_certs option
  * 1.1.11: Added support for TIFF
  * 1.1.12: Added alpha channel removing support
  * 1.1.13: Added original image size keeping


pilbox/app.py
#!/usr/bin/env python
#
# Copyright 2013 Adam Gschwender
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from __future__ import absolute_import, division, with_statement

import logging
import socket

import tornado.escape
import tornado.gen
import tornado.httpclient
import tornado.httpserver
import tornado.ioloop
import tornado.options
import tornado.web
from tornado.options import define, options, parse_config_file

from pilbox import errors
from pilbox.image import Image
from pilbox.signature import verify_signature

try:
    from urlparse import urlparse, urljoin
except ImportError:
    from urllib.parse import urlparse, urljoin

try:
    import pycurl
except ImportError:
    pycurl = None


# general settings
define("config", help="path to configuration file",
       callback=lambda path: parse_config_file(path, final=False))
define("debug", help="run in debug mode", type=bool, default=False)
define("port", help="run on the given port", type=int, default=8888)

# security related settings
define("client_name", help="client name")
define("client_key", help="client key")
define("allowed_hosts", help="valid hosts", default=[], multiple=True)
define("allowed_operations", help="valid ops", default=[], multiple=True)

# request related settings
define("max_requests", help="max concurrent requests", type=int, default=40)
define("timeout", help="request timeout in seconds", type=float, default=10)
define("implicit_base_url", help="prepend protocol/host to url paths")
define("ca_certs",
       help="override filename of CA certificates in PEM format",
       default=None)
define("validate_cert", help="validate certificates", type=bool, default=True)
define("proxy_host", help="proxy hostname")
define("proxy_port", help="proxy port", type=int)

# header related settings
define("content_type_from_image",
       help="override content type using image mime type",
       type=bool)

# default image option settings
define("background", help="default hexadecimal bg color (RGB or ARGB)")
define("remove_alpha", help="remove alpha changel from the source image", type=bool, default=False)
define("expand", help="default to expand when rotating", type=int)
define("filter", help="default filter to use when resizing")
define("format", help="default format to use when outputting")
define("mode", help="default mode to use when resizing")
define("operation", help="default operation to perform")
define("optimize", help="default to optimize when saving", type=int)
define("position", help="default cropping position")
define("progressive", help="default to progressive when saving", type=int)
define("quality", help="default jpeg quality, 1-99 or keep")
define("retain", help="default adaptive retain percent, 1-99", type=int)
define("preserve_exif", help="default behavior for exif data", type=int)

logger = logging.getLogger("tornado.application")


class PilboxApplication(tornado.web.Application):

    def __init__(self, **kwargs):
        settings = dict(
            debug=options.debug,
            client_name=options.client_name,
            client_key=options.client_key,
            allowed_hosts=options.allowed_hosts,
            allowed_operations=set(
                options.allowed_operations or ImageHandler.OPERATIONS),
            background=options.background,
            remove_alpha=options.remove_alpha,
            expand=options.expand,
            filter=options.filter,
            format=options.format,
            mode=options.mode,
            operation=options.operation,
            optimize=options.optimize,
            position=options.position,
            progressive=options.progressive,
            quality=options.quality,
            max_requests=options.max_requests,
            timeout=options.timeout,
            implicit_base_url=options.implicit_base_url,
            ca_certs=options.ca_certs,
            validate_cert=options.validate_cert,
            content_type_from_image=options.content_type_from_image,
            proxy_host=options.proxy_host,
            proxy_port=options.proxy_port,
            preserve_exif=options.preserve_exif)

        settings.update(kwargs)

        if settings.get("proxy_host") and pycurl is None:  # pragma: no cover
            raise Exception("PycURL is required for proxy requests")

        if pycurl is not None:  # pragma: no cover
            tornado.httpclient.AsyncHTTPClient.configure(
                "tornado.curl_httpclient.CurlAsyncHTTPClient")

        tornado.web.Application.__init__(self, self.get_handlers(), **settings)

    def get_handlers(self):
        return [(r"/", ImageHandler)]


class ImageHandler(tornado.web.RequestHandler):
    FORWARD_HEADERS = ["Cache-Control", "Expires", "Last-Modified"]
    OPERATIONS = ["region", "resize", "rotate", "noop"]

    _FORMAT_TO_MIME = {
        "gif": "image/gif",
        "jpeg": "image/jpeg",
        "jpg": "image/jpeg",
        "png": "image/png",
        "webp": "image/webp",
        "tiff": "image/tiff",
    }

    @tornado.gen.coroutine
    def get(self):
        self.validate_request()
        resp = yield self.fetch_image()
        self.render_image(resp)

    def get_argument(self, name, default=None, strip=True):
        return super(ImageHandler, self).get_argument(name, default, strip)

    def validate_request(self):
        self._validate_operation()
        self._validate_url()
        self._validate_signature()
        self._validate_client()
        self._validate_host()

        opts = self._get_save_options()
        ops = self._get_operations()
        if "resize" in ops:
            Image.validate_dimensions(
                self.get_argument("w"), self.get_argument("h"))
            opts.update(self._get_resize_options())
        if "rotate" in ops:
            Image.validate_degree(self.get_argument("deg"))
            opts.update(self._get_rotate_options())
        if "region" in ops:
            Image.validate_rectangle(self.get_argument("rect"))

        Image.validate_options(opts)

    @tornado.gen.coroutine
    def fetch_image(self):
        url = self.get_argument("url")
        if self.settings.get("implicit_base_url") \
                and urlparse(url).hostname is None:
            url = urljoin(self.settings.get("implicit_base_url"), url)

        client = tornado.httpclient.AsyncHTTPClient(
            max_clients=self.settings.get("max_requests"))
        try:
            resp = yield client.fetch(
                url,
                request_timeout=self.settings.get("timeout"),
                ca_certs=self.settings.get("ca_certs"),
                validate_cert=self.settings.get("validate_cert"),
                proxy_host=self.settings.get("proxy_host"),
                proxy_port=self.settings.get("proxy_port"))
            raise tornado.gen.Return(resp)
        except (socket.gaierror, tornado.httpclient.HTTPError) as e:
            logger.warn("Fetch error for %s: %s",
                        self.get_argument("url"),
                        str(e))
            raise errors.FetchError()

    def render_image(self, resp):
        outfile, outfile_format = self._process_response(resp)
        self._set_headers(resp.headers, outfile_format)
        for block in iter(lambda: outfile.read(65536), b""):
            self.write(block)
        outfile.close()

    def write_error(self, status_code, **kwargs):
        err = kwargs["exc_info"][1] if "exc_info" in kwargs else None
        if isinstance(err, errors.PilboxError):
            self.set_header("Content-Type", "application/json")
            resp = dict(status_code=status_code,
                        error_code=err.get_code(),
                        error=err.log_message)
            self.finish(tornado.escape.json_encode(resp))
        else:
            super(ImageHandler, self).write_error(status_code, **kwargs)

    def _process_response(self, resp):
        ops = self._get_operations()
        if "noop" in ops:
            return (resp.buffer, None)

        image = Image(resp.buffer)
        self._image_remove_alpha(image)
        for operation in ops:
            if operation == "resize":
                self._image_resize(image)
            elif operation == "rotate":
                self._image_rotate(image)
            elif operation == "region":
                self._image_region(image)

        return (self._image_save(image), image.img.format)

    def _image_region(self, image):
        image.region(self.get_argument("rect").split(","))

    def _image_resize(self, image):
        opts = self._get_resize_options()
        image.resize(self.get_argument("w"), self.get_argument("h"), **opts)

    def _image_remove_alpha(self, image):
        opts = self._get_remove_alpha_options()
        if opts['remove_alpha']:
            image._remove_alpha_channel(background=opts['background'])

    def _image_rotate(self, image):
        opts = self._get_rotate_options()
        image.rotate(self.get_argument("deg"), **opts)

    def _image_save(self, image):
        opts = self._get_save_options()
        return image.save(**opts)

    def _set_headers(self, headers, file_format):
        if file_format and any((self.get_argument("fmt"),
                                self.settings.get("format"),
                                self.settings.get("content_type_from_image"))):
            self.set_header(
                "Content-Type", self._FORMAT_TO_MIME.get(file_format.lower()))
        elif "Content-Type" in headers:
            self.set_header("Content-Type", headers["Content-Type"])

        for k in ImageHandler.FORWARD_HEADERS:
            if k in headers and headers[k]:
                self.set_header(k, headers[k])

    def _get_operations(self):
        return self.get_argument(
            "op", self.settings.get("operation") or "resize").split(",")

    def _get_resize_options(self):
        return self._get_options(
            dict(mode=self.get_argument("mode"),
                 filter=self.get_argument("filter"),
                 position=self.get_argument("pos"),
                 background=self.get_argument("bg"),
                 retain=self.get_argument("retain")))

    def _get_rotate_options(self):
        return self._get_options(
            dict(expand=self.get_argument("expand")))

    def _get_remove_alpha_options(self):
        return self._get_options(dict(background=self.get_argument("bg"),
                                      remove_alpha=None))

    def _get_save_options(self):
        return self._get_options(
            dict(format=self.get_argument("fmt"),
                 optimize=self.get_argument("opt"),
                 quality=self.get_argument("q"),
                 progressive=self.get_argument("prog"),
                 preserve_exif=self.get_argument("exif")))

    def _get_options(self, opts):
        for k, v in opts.items():
            if v is None:
                opts[k] = self.settings.get(k, None)
        return opts

    def _validate_operation(self):
        operations = set(self._get_operations())
        if not operations.issubset(self.settings.get("allowed_operations")):
            raise errors.OperationError("Unsupported operation")

    def _validate_url(self):
        url = self.get_argument("url")
        if not url:
            raise errors.UrlError("Missing url")
        elif url.startswith("http://") or url.startswith("https://"):
            return
        elif self.settings.get("implicit_base_url") and url.startswith("/"):
            return
        raise errors.UrlError("Unsupported protocol")

    def _validate_client(self):
        client = self.settings.get("client_name")
        if client and self.get_argument("client") != client:
            raise errors.ClientError("Invalid client")

    def _validate_signature(self):
        key = self.settings.get("client_key")
        if key and not verify_signature(key, urlparse(self.request.uri).query):
            raise errors.SignatureError("Invalid signature")

    def _validate_host(self):
        hosts = self.settings.get("allowed_hosts", [])
        if hosts and urlparse(self.get_argument("url")).hostname not in hosts:
            raise errors.HostError("Invalid host")


def parse_command_line():  # pragma: no cover
    tornado.options.parse_command_line()


def start_server(app=None):  # pragma: no cover
    if options.debug:
        logger.setLevel(logging.DEBUG)
    server = tornado.httpserver.HTTPServer(
        app if app else PilboxApplication())
    logger.info("Starting server...")
    try:
        server.bind(options.port)
        server.start(1 if options.debug else 0)
        tornado.ioloop.IOLoop.instance().start()
    except KeyboardInterrupt:
        tornado.ioloop.IOLoop.instance().stop()


def main(app=None):
    parse_command_line()
    start_server(app)


if __name__ == "__main__":
    main()
